js/src/Ice/ObjectPrxExtensions.js

```diff
@@ -366,113 +360,33 @@ ObjectPrx.prototype._getReference = function () {
-    // can safely be repeated.
-    //
-    // An ObjectNotExistException can always be retried as well without violating
-    // "at-most-once" (see the implementation of the checkRetryAfterException method
-    //  of the ProxyFactory class for the reasons why it can be useful).
-    //
-    // If the request didn't get sent or if it's non-mutating or idempotent it can
-    // also always be retried if the retry count isn't reached.
-    //
-    if (
-        ex instanceof LocalException &&
-        (!sent ||
-            mode == OperationMode.Nonmutating ||
-            mode == OperationMode.Idempotent ||
-            ex instanceof CloseConnectionException ||
-            ex instanceof ObjectNotExistException)
-    ) {
-        try {
-            return this._reference
-                .getInstance()
-                .proxyFactory()
-                .checkRetryAfterException(ex, this._reference, sleep, cnt);
-        } catch (exc) {
-            if (exc instanceof CommunicatorDestroyedException) {
-                //
-                // The communicator is already destroyed, so we cannot retry.
-                //
-                throw ex;
-            } else {
-                throw exc;
-            }
-        }
-    } else {
-        throw ex;
-    }
-};
-
 ObjectPrx.prototype._checkAsyncTwowayOnly = function (name) {
     if (!this.ice_isTwoway()) {
         throw new TwowayOnlyException(name);
     }
 };
 
-ObjectPrx.prototype._getRequestHandler = function () {
-    if (this._reference.getCacheConnection()) {
-        if (this._requestHandler) {
-            return this._requestHandler;
-        }
-    }
-    return this._reference.getRequestHandler(this);
-};
-
-ObjectPrx.prototype._getBatchRequestQueue = function () {
-    if (!this._batchRequestQueue) {
-        this._batchRequestQueue = this._reference.getBatchRequestQueue();
-    }
-    return this._batchRequestQueue;
-};
-
-ObjectPrx.prototype._setRequestHandler = function (handler) {
-    if (this._reference.getCacheConnection()) {
-        if (!this._requestHandler) {
-            this._requestHandler = handler;
-        }
-        return this._requestHandler;
-    }
-    return handler;
-};
-
-ObjectPrx.prototype._updateRequestHandler = function (previous, handler) {
-    if (this._reference.getCacheConnection() && previous !== null) {
-        if (this._requestHandler && this._requestHandler !== handler) {
-            this._requestHandler = this._requestHandler.update(previous, handler);
-        }
-    }
-};
-
 //
-// Only for use by IceInternal.ProxyFactory
+// Only for use by ObjectPrx constructor
 //
-ObjectPrx.prototype._setup = function (ref) {
-    Debug.assert(this._reference === null);
-
-    this._reference = ref;
+ObjectPrx.prototype._setup = function (arg0, proxyString = "") {
+    Debug.assert(this._reference === undefined);
+    if (arg0 instanceof Communicator) {
+        this._reference = arg0.instance.referenceFactory().createFromString(proxyString, "");
```

Comment:
We should throw if _reference is null (because proxyString is empty).

Reply:
fixed